firedrake/eigensolver.py

```diff
@@ -0,0 +1,192 @@
+"""Specify and solve finite element eigenproblems."""
+from firedrake.assemble import assemble
+from firedrake.function import Function
+from firedrake import utils
+from firedrake.petsc import OptionsManager, flatten_parameters
+from firedrake.exceptions import ConvergenceError
+try:
+    from slepc4py import SLEPc
+except ImportError:
+    SLEPc = None
+__all__ = ["LinearEigenproblem",
+           "LinearEigensolver"]
+
+
+class LinearEigenproblem():
+    """Generalised linear eigenvalue problem.
+
+    The problem has the form, find `u`, `λ` such that::
+
+        A(u, v) = λM(u, v)    ∀ v ∈ V
+
+    Parameters
+    ----------
+    A : ufl.Form
+        The bilinear form A(u, v).
+    M : ufl.Form
+        The mass form M(u, v),  defaults to inner(u, v) * dx.
+    bcs : DirichletBC or list of DirichletBC
+        The boundary conditions.
+    bc_shift: float
+        The value to shift the boundary condition eigenvalues by.
+
+    Notes
+    -----
+
+    If Dirichlet boundary conditions are supplied then these will result in the
+    eigenproblem having a nullspace spanned by the basis functions with support
+    on the boundary. To facilitate solution, this is shifted by the specified
+    amount. It is the user's responsibility to ensure that the shift is not
+    close to an actual eigenvalue of the system.
+    """
+    def __init__(self, A, M=None, bcs=None, bc_shift=666.0):
+        if not SLEPc:
+            raise ImportError(
+                "Unable to import SLEPc, eigenvalue computation not possible "
+                "(try firedrake-update --slepc)"
+            )
+
+        self.A = A  # LHS
+        args = A.arguments()
+        v, u = args
+        if M:
+            self.M = M
+        else:
+            from ufl import inner, dx
+            self.M = inner(u, v) * dx
```

Comment:
This is very specific. One might be interested in using the eigensolver to compute the eigenvalue of a saddle point problem, which usually has a different mass matrix with a 0 block on the diagonal.

Reply:
It's only a default, and one that is correct in the straightforward case. The user can always provide a different `M`.